cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +423,303 @@ class self_comparator {
+            if (thrust::any_of(
+                  thrust::seq, lvalid, lvalid + lcol.size(), thrust::logical_not<bool>()) or
+                thrust::any_of(
+                  thrust::seq, rvalid, rvalid + rcol.size(), thrust::logical_not<bool>())) {
+              return false;
+            }
+          } else {
+            if (not thrust::equal(thrust::seq, lvalid, lvalid + lcol.size(), rvalid)) {
+              return false;
+            }
+          }
+        }
+        if (lcol.type().id() == type_id::STRUCT) {
+          lcol = detail::structs_column_device_view(lcol).sliced_child(0);
+          rcol = detail::structs_column_device_view(rcol).sliced_child(0);
+        } else if (lcol.type().id() == type_id::LIST) {
+          auto l_list_col = detail::lists_column_device_view(lcol);
+          auto r_list_col = detail::lists_column_device_view(rcol);
+
+          auto lsizes = make_list_size_iterator(l_list_col);
+          auto rsizes = make_list_size_iterator(r_list_col);
+          if (not thrust::equal(thrust::seq, lsizes, lsizes + lcol.size(), rsizes)) {
+            return false;
+          }
+
+          lcol = l_list_col.sliced_child();
+          rcol = r_list_col.sliced_child();
+          if (lcol.size() != rcol.size()) { return false; }
+        }
+      }
+
+      auto comp = element_range_comparator{element_comparator{nulls, lcol, rcol, nulls_are_equal}};
+      return type_dispatcher<dispatch_void_if_nested>(lcol.type(), comp, 0, 0, lcol.size());
+    }
+
+   private:
+    /**
+     * @brief Compares a range of elements in two columns for equality.
+     *
+     * When we want to compare a range of elements in a single column with an equal sized range of
+     * elements from another column, this saves us from type dispatching for each individual element
+     * in the range
+     */
+    struct element_range_comparator {
+      element_comparator const comp;
+
+      /**
+       * @brief Compare a range of elements for equality.
+       *
+       * Given two columns, compare the elements in the range
+       * [lhs_start_index, lhs_start_index + size) in lhs column with the elements in the range
+       * [rhs_start_index, rhs_start_index + size) in rhs column.
+       * `lhs` and `rhs` columns are defined as the first and second columns used to construct
+       * member `element_comparator comp`
+       *
+       * @param lhs_start_index Starting index of the range of elements in the lhs column
+       * @param rhs_start_index Starting index of the range of elements in the rhs column
+       * @param size The number of elements in the range
+       * @return True if ALL elements in the range compare equal, false otherwise
+       */
+      template <typename Element, CUDF_ENABLE_IF(cudf::is_equality_comparable<Element, Element>())>
+      __device__ bool operator()(size_type const lhs_start_index,
+                                 size_type const rhs_start_index,
```

Comment:
> What's the use-case for having two different start indexes? That seems like a dangerous footgun for out of bounds accesses, especially since the only use case that we appear to have here is for both to start at 0.

Out of bounds access won't be prevented by having the same starting index. We could go out of bounds for both.

> Is there a sliced case where we might want to operator on different offsets of a column?

Not atm but then it would become a very specialized class that won't be useful for anything else.

Reply:
How about this https://github.com/rapidsai/cudf/pull/10289/commits/5bbf18e65f0fcfc2958b4e61b4012bfdd669c414